cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/DatasetInstanceService.java

```diff
@@ -0,0 +1,302 @@
+    if (spec == null) {
+      throw new NotFoundException(instance);
+    }
+
+    Id.DatasetType datasetTypeId = Id.DatasetType.from(instance.getNamespace(), spec.getType());
+    DatasetTypeMeta typeMeta = getTypeInfo(instance.getNamespace(), spec.getType());
+    if (typeMeta == null) {
+      // TODO: This shouldn't happen unless CDAP is in an invalid state - maybe give different error
+      throw new NotFoundException(datasetTypeId);
+    }
+
+    registerUsage(instance, owners);
+    return new DatasetMeta(spec, typeMeta, null);
+  }
+
+  private void registerUsage(Id.DatasetInstance instance, List<? extends Id> owners) {
+    for (Id owner : owners) {
+      try {
+        if (owner instanceof Id.Program) {
+          usageRegistry.register((Id.Program) owner, instance);
+        } else if (owner instanceof Id.Adapter) {
+          usageRegistry.register((Id.Adapter) owner, instance);
+        }
+      } catch (Exception e) {
+        LOG.warn("Failed to register usage of {} -> {}", owner, instance);
+      }
+    }
+  }
+
+  /**
+   * Creates a dataset instance.
+   *
+   * @param namespace the namespace to create the dataset instance in
+   * @param name the name of the new dataset instance
+   * @param props the properties for the new dataset instance
+   * @throws DatasetAlreadyExistsException if a dataset with the same name already exists
+   * @throws DatasetTypeNotFoundException if the dataset type was not found
+   * @throws Exception if something went wrong
+   */
+  public void create(Id.Namespace namespace, String name, DatasetInstanceConfiguration props) throws Exception {
+    Id.DatasetInstance newInstance = Id.DatasetInstance.from(namespace, name);
+    DatasetSpecification existing = instanceManager.get(newInstance);
+    if (existing != null && !allowDatasetUncheckedUpgrade) {
+      throw new DatasetAlreadyExistsException(newInstance);
+    }
+
+    // Disable explore if the table already existed
+    if (existing != null) {
+      disableExplore(newInstance);
+    }
+
+    DatasetTypeMeta typeMeta = getTypeInfo(namespace, props.getTypeName());
+    if (typeMeta == null) {
+      // Type not found in the instance's namespace and the system namespace. Bail out.
+      throw new DatasetTypeNotFoundException(Id.DatasetType.from(namespace, props.getTypeName()));
+    }
+
+    // Note how we execute configure() via opExecutorClient (outside of ds service) to isolate running user code
+    DatasetSpecification spec = opExecutorClient.create(newInstance, typeMeta,
+                                                        DatasetProperties.builder()
+                                                          .addAll(props.getProperties())
+                                                          .build());
+    instanceManager.add(namespace, spec);
+
+    // Enable explore
+    enableExplore(newInstance, props);
+  }
+
+  /**
+   * Updates an existing Dataset specification properties  {@link DatasetInstanceConfiguration}
```

Comment:
Missing a period. Should be "Updates an existing Dataset specification properties."